File: insta/app/src/main/java/com/example/insta/NewsListAdapter.java
package com.example.insta;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.List;

public class NewsListAdapter extends RecyclerView.Adapter<NewsListAdapter.NewsViewHolder> {

    private List<News> newsList;
    @Nullable
    private ItemClickListener listener;

    public NewsListAdapter(List<News> newsList) {
        this.newsList = newsList;
    }
    public NewsListAdapter(List<News> newsList, @Nullable ItemClickListener listener) {
        this.newsList = newsList;
        this.listener = listener;
    }

    @NonNull
    @Override
    public NewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_news, null, false);
        RecyclerView.LayoutParams params = new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        view.setLayoutParams(params);
        return new NewsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsViewHolder holder, final int position) {
        final News news = newsList.get(position);
        holder.tvAuthor.setText(Html.fromHtml(news.getAuthor()));
        holder.tvTitle.setText(Html.fromHtml(news.getTitle()));
        holder.tvComments.setText(String.valueOf(news.getCommentsCount()));
        holder.tvViews.setText(news.getViewsCount());
        Picasso.get().load(news.getAva()).into(holder.ava);
        Picasso.get().load(news.getPhotoLink()).into(holder.postPhoto);
        holder.postData.setText(news.getPostData());

        holder.buttonLike.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.onClick(v, position, news);
                }
            }
        });

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (listener != null) {

                    listener.itemClick(position, news);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return newsList.size();
    }

    public class NewsViewHolder extends RecyclerView.ViewHolder {

        TextView tvTitle;
        TextView tvAuthor;
        TextView tvViews;
        TextView tvComments;
        ImageView ava;
        ImageView postPhoto;
        TextView postData;

        Button buttonLike;
        Button moreText;

        public NewsViewHolder(@NonNull View itemView) {
            super(itemView);
            buttonLike = (Button)itemView.findViewById(R.id.postLike);


            ava = (ImageView)itemView.findViewById(R.id.imageView3);
            postPhoto = (ImageView) itemView.findViewById(R.id.imageButton);

            postData = itemView.findViewById(R.id.postData);

            tvTitle = itemView.findViewById(R.id.tvTitle);
            tvAuthor = itemView.findViewById(R.id.tvDate);

            tvViews = itemView.findViewById(R.id.tvViews);
            tvComments = itemView.findViewById(R.id.tvComments);
        }


    }

    public interface ItemClickListener {
        public void itemClick(int position, News item);
        public void onClick(View v, int position, News item);
    }
}

File: insta/app/src/main/java/com/example/insta/NewsListAdapter.kt
package com.example.insta

import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.squareup.picasso.Picasso
class NewsListAdapter : RecyclerView.Adapter<NewsListAdapter.NewsViewHolder> {

    private var newsList: List<News>? = null
    private var listener: ItemClickListener?

    constructor(newsList: List<News>, listener: ItemClickListener?) {
        this.newsList = newsList
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_news, null, false)
        val params = RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        )
        view.layoutParams = params
        return NewsViewHolder(view)
    }

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        val news = newsList!![position]
        holder.tvAuthor.text = Html.fromHtml(news.author)
        holder.tvTitle.text = Html.fromHtml(news.title)
        holder.tvComments.text = news.commentsCount.toString()
        holder.tvViews.text = news.viewsCount
        Picasso.get().load(news.ava).into(holder.ava)
        Picasso.get().load(news.photoLink).into(holder.postPhoto)
        holder.postData.text = news.postData

        holder.buttonLike.setOnClickListener { v ->
            listener?.onClick(v, position, news)
        }

        holder.itemView.setOnClickListener {
            listener?.itemClick(position, news)
        }
    }

    override fun getItemCount(): Int {
        return newsList!!.size
    }

    inner class NewsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        internal var tvTitle: TextView
        internal var tvAuthor: TextView
        internal var tvViews: TextView
        internal var tvComments: TextView
        internal var ava: ImageView
        internal var postPhoto: ImageView
        internal var postData: TextView

        internal var buttonLike: Button
        internal var moreText: Button? = null

        init {
            buttonLike = itemView.findViewById<View>(R.id.postLike) as Button


            ava = itemView.findViewById<View>(R.id.imageView3) as ImageView
            postPhoto = itemView.findViewById<View>(R.id.imageButton) as ImageView

            postData = itemView.findViewById(R.id.postData)

            tvTitle = itemView.findViewById(R.id.tvTitle)
            tvAuthor = itemView.findViewById(R.id.tvDate)

            tvViews = itemView.findViewById(R.id.tvViews)
            tvComments = itemView.findViewById(R.id.tvComments)
        }


    }

    interface ItemClickListener {
        fun itemClick(position: Int, item: News)
        fun onClick(v: View, position: Int, item: News)
    }
}
